File: docs/chapters/18-event-driven-architecture/go/transactional_outbox.go
// Scenario: checkout commits Order + outbox row; worker publishes asynchronously
// Demonstrates: transactional outbox in one DB transaction
// Trade-off: publish lag acceptable for async fulfillment; not for sync read-your-writes

package main

import (
	"context"
	"database/sql"
	"encoding/json"
)

type OutboxRow struct {
	EventID string
	Topic   string
	Payload []byte
}

func placeOrder(ctx context.Context, tx *sql.Tx, orderID, userID string) error {
	if _, err := tx.ExecContext(ctx,
		`INSERT INTO orders (id, user_id, status) VALUES ($1, $2, 'PLACED')`,
		orderID, userID); err != nil {
		return err
	}
	payload, _ := json.Marshal(map[string]string{
		"type": "OrderPlaced", "orderId": orderID,
	})
	_, err := tx.ExecContext(ctx,
		`INSERT INTO outbox (event_id, topic, payload, published) VALUES (gen_random_uuid(), $1, $2, false)`,
		"order.events", payload)
	return err
}

File: chapters/18-event-driven-architecture/java/TransactionalOutbox.java
// Scenario: checkout commits Order + outbox row; separate worker publishes to bus
// Demonstrates: transactional outbox — no dual-write to DB and broker in one request
// Trade-off: seconds of publish lag vs lost events on crash-after-commit

import java.sql.Connection;
import java.util.UUID;

public class TransactionalOutbox {

    public void placeOrder(Connection conn, String orderId, String userId) throws Exception {
        conn.setAutoCommit(false);
        try {
            try (var ps = conn.prepareStatement(
                    "INSERT INTO orders (id, user_id, status) VALUES (?, ?, 'PLACED')")) {
                ps.setString(1, orderId);
                ps.setString(2, userId);
                ps.executeUpdate();
            }
            String eventId = UUID.randomUUID().toString();
            try (var ps = conn.prepareStatement(
                    "INSERT INTO outbox (event_id, topic, payload, published) VALUES (?, ?, ?, false)")) {
                ps.setString(1, eventId);
                ps.setString(2, "order.events");
                ps.setString(3, "{\"type\":\"OrderPlaced\",\"orderId\":\"" + orderId + "\"}");
                ps.executeUpdate();
            }
            conn.commit();
        } catch (Exception e) {
            conn.rollback();
            throw e;
        }
    }
}
